src/devices/machine/swtpc8212.cpp

```diff
@@ -0,0 +1,475 @@
+	PORT_DIPSETTING(0x1c, "19200")
+	PORT_DIPSETTING(0x1f, "38400")
+	PORT_DIPNAME(0x20, 0x00, "Mode switch") PORT_DIPLOCATION("DIP:5")
+	PORT_DIPSETTING(0x00, "Conversational")
+	PORT_DIPSETTING(0x20, "Page edit")
+	PORT_DIPNAME(0x40, 0x00, "No Parity") PORT_DIPLOCATION("DIP:6")
+	PORT_DIPSETTING(0x00, "No Parity")
+	PORT_DIPSETTING(0x40, "Parity")
+	PORT_DIPNAME(0x80, 0x00, "Parity Select") PORT_DIPLOCATION("DIP:7")
+	PORT_DIPSETTING(0x00, "Odd or Mark")
+	PORT_DIPSETTING(0x80, "Even or Space")
+
+	PORT_START("config")
+	PORT_CONFNAME(0x01, 0x01, "Duplex")
+	PORT_CONFSETTING(0x00, "Full duplex")
+	PORT_CONFSETTING(0x01, "Half duplex")
+	PORT_CONFNAME(0x02, 0x00, "Option ROM (Not used)")
+	PORT_CONFSETTING(0x00, DEF_STR(On))
+	PORT_CONFSETTING(0x02, DEF_STR(Off))
+	PORT_CONFNAME(0x04, 0x04, "Parity Select (Jumper B)")
+	PORT_CONFSETTING(0x00, "Odd or Even (On)")
+	PORT_CONFSETTING(0x04, "Mark or Space (Off)")
+	PORT_CONFNAME(0x08, 0x08, "Data bits (Jumper A)")
+	PORT_CONFSETTING(0x00, "7 bit data (On)")
+	PORT_CONFSETTING(0x08, "8 bit data (Off)")
+
+	PORT_START("one_stop_bit")
+	PORT_CONFNAME(1, 1, "One stop bit patch")
+	PORT_CONFSETTING(0, "No")
+	PORT_CONFSETTING(1, "Yes - apply patch")
+
+INPUT_PORTS_END
+
+
+ROM_START( swtpc8212 )
+	ROM_REGION( 0x1000, "program", 0 )
+	ROM_LOAD( "8224g_ver.1.1_6oct80.ic1", 0x0000, 0x0800, CRC(7d7f3c21) SHA1(f7e6e20b36a1c724a4e348bc784d0b7b5fb462a3) )
+	ROM_LOAD( "8224g_ver.1.1_6oct80.ic2", 0x0800, 0x0800, CRC(2b118c22) SHA1(5fa031c834c7c582d5715764941499fcef51f477) )
+
+	ROM_REGION( 0x0800, "chargen1", 0 )
+	ROM_LOAD( "mcm66750.rom",  0x0000, 0x0800, CRC(aedc2830) SHA1(49ce17d5b5cefb24e89ed3fd59887a652501b919) )
+	ROM_REGION( 0x0800, "chargen2", 0 )
+	ROM_LOAD( "grafix_8x12_22aug80.bin",  0x0000, 0x0800, CRC(a525ed65) SHA1(813d2e85ddb258c5b032b959e695ad33200cbcc4) )
+ROM_END
+
+void swtpc8212_device::device_add_mconfig(machine_config &config)
+{
+	M6802(config, m_maincpu, 1.8432_MHz_XTAL);
+	m_maincpu->set_addrmap(AS_PROGRAM, &swtpc8212_device::mem_map);
+
+	INPUT_MERGER_ANY_HIGH(config, "mainirq").output_handler().set_inputline(m_maincpu, M6802_IRQ_LINE);
+
+	// PA - various jumpers and cursor inputs.
+	// PB - parallel keyboard data input.
+	// CA1 - cursor input strobe?
+	// CA2 - output, bell.
+	// CB1 - parallel keyboard input strobe.
+	// CB2 ??
+	pia6821_device &pia0(PIA6821(config, "pia0"));
+	pia0.readpa_handler().set(FUNC(swtpc8212_device::pia0_pa_r));
+	pia0.readpb_handler().set(FUNC(swtpc8212_device::pia0_pb_r));
+	pia0.ca2_handler().set(FUNC(swtpc8212_device::pia0_ca2_w));
+	pia0.irqa_handler().set("mainirq", FUNC(input_merger_device::in_w<0>));
+	pia0.irqb_handler().set("mainirq", FUNC(input_merger_device::in_w<1>));
+
+	// PA - parallel printer data outputs.
+	// PB - various config inputs.
+	// CA1 - parallel printer, 'busy' input.
+	// CA2 - parallel printer, 'data ready' output.
+	// CB1 - Handshake input?
+	// CB2 - Handshake output?
+	pia6821_device &pia1(PIA6821(config, "pia1"));
```

Comment:
When you have a device finder, you should use it in place of the literal tag and avoid the need for a temporary when building machine config.  It's less error-prone.  This applies to all the devices here that you have device finders for.